evaluation.py

```diff
@@ -237,7 +237,97 @@ def predict_keypoints(self,generator, location):
+            "right_hip": 0,
+            "left_knee": 0,
+            "right_knee": 0,
+            "left_ankle": 0,
+            "right_ankle": 0
+        }
+
+        for prediction in list_of_predictions:
+            prediction_image_id = prediction['image_id']
+            prediction_keypoints = prediction['keypoints']
+            for i, ann in enumerate(data['annotations']):
+                if data['annotations'][i]['image_id'] == prediction_image_id:
+                    annotation_keypoints = data['annotations'][i]['keypoints']
+                    prediction_keypoints = np.array(prediction_keypoints)
+                    annotation_keypoints = np.array(annotation_keypoints)
+
+                    # Joint at 11 is left hip, Joint at 12 is right hip
+                    left_hip_point = np.array(annotation_keypoints[33], annotation_keypoints[34])
+                    right_hip_point = np.array(annotation_keypoints[36], annotation_keypoints[37])
+                    # PCK@0.2 == Distance between predicted and true joint < 0.2 * torso diameter
+                    torso = np.linalg.norm(left_hip_point-right_hip_point)
+                    threshold = 0.2*torso
+
+                    for i in range(50):
+                        if i % 3 == 0:
+                            prediction_point = np.array(prediction_keypoints[i], prediction_keypoints[i+1])
+                            annotation_point = np.array(annotation_keypoints[i], annotation_keypoints[i+1])
+                            dist = (np.linalg.norm(prediction_point-annotation_point))
+                            dist_list.append(dist)
+
+                    correct_keypoints["nose"] += dist_list[0]<threshold
+                    correct_keypoints["left_eye"] += dist_list[1]<threshold
+                    correct_keypoints["right_eye"] += dist_list[2]<threshold
+                    correct_keypoints["left_ear"] += dist_list[3]<threshold
+                    correct_keypoints["right_ear"] += dist_list[4]<threshold
+                    correct_keypoints["left_shoulder"] += dist_list[5]<threshold
+                    correct_keypoints["right_shoulder"] += dist_list[6]<threshold
+                    correct_keypoints["left_elbow"] += dist_list[7]<threshold
+                    correct_keypoints["right_elbow"] += dist_list[8]<threshold
+                    correct_keypoints["left_wrist"] += dist_list[9]<threshold
+                    correct_keypoints["right_wrist"] += dist_list[10]<threshold
+                    correct_keypoints["left_hip"] += dist_list[11]<threshold
+                    correct_keypoints["right_hip"] += dist_list[12]<threshold
+                    correct_keypoints["left_knee"] += dist_list[13]<threshold
+                    correct_keypoints["right_knee"] += dist_list[14]<threshold
+                    correct_keypoints["left_ankle"] += dist_list[15]<threshold
+                    correct_keypoints["right_ankle"] += dist_list[16]<threshold
```

Comment:
For PCK, do we count the model predicting **no joint** as success? (I feel like you've answered this before but I can't remember)

Reply:
So, yes it will see how far away it was from the true point. So if our model predicts 0, 0 and the ground truth is 0, 0 it will definitely be under the threshold and count as a correct point.

Reply:
That sounds good! Is the absence of a joint encoded as 0,0 for both ground truth and predicted?

Reply:
Yep it should be.